src/llfs/no_outgoing_refs_cache.cpp

```diff
@@ -0,0 +1,75 @@
+//#=##=##=#==#=#==#===#+==#+==========+==+=+=+=+=+=++=+++=+++++=-++++=-+++++++++++
+//
+// Part of the LLFS Project, under Apache License v2.0.
+// See https://www.apache.org/licenses/LICENSE-2.0 for license information.
+// SPDX short identifier: Apache-2.0
+//
+//+++++++++++-+-+--+----- --- -- -  -  -   -
+#include <llfs/no_outgoing_refs_cache.hpp>
+
+namespace llfs {
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+NoOutgoingRefsCache::NoOutgoingRefsCache(const PageIdFactory& page_ids) noexcept
+    : page_ids_{page_ids}
+    , cache_(this->page_ids_.get_physical_page_count().value())
+{
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+void NoOutgoingRefsCache::set_page_state(PageId page_id,
+                                         batt::BoolStatus new_has_outgoing_refs_state) noexcept
+{
+  BATT_CHECK_EQ(PageIdFactory::get_device_id(page_id), this->page_ids_.get_device_id());
+  const u64 physical_page = this->page_ids_.get_physical_page(page_id);
+  const page_generation_int generation = this->page_ids_.get_generation(page_id);
+  BATT_CHECK_LT((usize)physical_page, this->cache_.size());
+
+  u64 new_cache_entry = generation << this->kGenerationShift;
+
+  // Set the "valid" bit to 1.
+  //
+  new_cache_entry |= this->kValidBitMask;
+
+  // If new_has_outgoing_refs_state has value kFalse, page_id has no outgoing references.
+  //
+  if (new_has_outgoing_refs_state == batt::BoolStatus::kFalse) {
+    // Set the "has no outgoing references" bit to 1.
+    //
+    new_cache_entry |= u64{1};
+  }
+
+  u64 old_cache_entry =
+      this->cache_[physical_page].exchange(new_cache_entry, std::memory_order_acq_rel);
+
+  // Sanity check: we are not setting the bits for the same generation more than once.
+  //
+  page_generation_int old_generation = old_cache_entry >> this->kGenerationShift;
+  BATT_CHECK_NE(generation, old_generation);
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+batt::BoolStatus NoOutgoingRefsCache::has_outgoing_refs(PageId page_id) const noexcept
+{
+  BATT_CHECK_EQ(PageIdFactory::get_device_id(page_id), this->page_ids_.get_device_id());
+  const u64 physical_page = this->page_ids_.get_physical_page(page_id);
+  const page_generation_int generation = this->page_ids_.get_generation(page_id);
+  BATT_CHECK_LT((usize)physical_page, this->cache_.size());
+
+  u64 current_cache_entry = this->cache_[physical_page].load(std::memory_order_acquire);
+  page_generation_int stored_generation = current_cache_entry >> this->kGenerationShift;
+  OutgoingRefsStatus outgoing_refs_status =
```

Comment:
We can get rid of the enum `OutgoingRefsStatus` here, and just use a switch to handle all the possible cases for the 2 least-significant bits.